database/migrations/2020_06_15_194355_create_insegnamento_table.php

```diff
@@ -20,114 +20,119 @@ class CreateInsegnamentoTable extends Migration
-            $table->string('canale')
+            $table->string('canale', 6)
                 ->nullable()
-                ->default(''); 
+                ->default('');
 
             /**
-             * Nel caso l'insegnamento disponga di più moduli, 
+             * Nel caso l'insegnamento disponga di più moduli,
              * si riporta l'identificativo di riferimento
              */
             $table->string('id_modulo')
                 ->nullable()
-                ->default(null); 
-            
+                ->default(null);
+
             /**
-             * Nel caso l'insegnamento disponga di più moduli, 
+             * Nel caso l'insegnamento disponga di più moduli,
              * si riporta il nome modulo di riferimento
              */
-            $table->string('nome_modulo')
+            $table->string('nome_modulo', 100)
                 ->nullable()
-                ->default(null); 
-            
+                ->default(null);
+
             /**
              * ??
              */
-            $table->string('tipo')
+            $table->string('tipo', 4)
                 ->nullable()
-                ->default(NULL); 
-            
+                ->default(NULL);
+
             /**
-             * campo di ricerca 
+             * campo di ricerca
              */
-            $table->string('ssd')
+            $table->string('ssd', 100)
                 ->nullable()
-                ->default(NULL); 
+                ->default(NULL);
 
             /**
              * ??
              */
-            $table->string('assegn')
+            $table->string('assegn', 100)
                 ->nullable()
-                ->default(NULL); 
+                ->default(NULL);
 
             /**
              * Identificativo del corso di studi di appartenenza
              */
-            $table->unsignedBigInteger('id_cds'); 
+            $table->unsignedBigInteger('id_cds');
 
-            $table->foreign('id_cds')
-                ->references('id')->on('corso_di_studi');
+            // $table->foreign('id_cds')
+            //     ->references('id')->on('corso_di_studi');
 
             /**
-             * L'identificativo fornito dall'unict, insieme 
+             * L'identificativo fornito dall'unict, insieme
              * all'anno accademico, forniscono una chiave univoca
              * abbastanza scomoda da utilizzare. Tuttavia, è
-             * necessario garantirne l'univocità. 
+             * necessario garantirne l'univocità.
              */
             $table->unique([
-                'codice_gomp', 
-                'anno_accademico', 
-                'id_modulo', 
-                'canale'
-            ]);
+                'codice_gomp',
+                'anno_accademico',
+                'id_modulo',
+                'canale',
+                'id_cds',
+                'anno',
+                'docente',
+                'assegn',
+                'tipo',
+            ], 'unique_ins');
```

Comment:
Ed ecco perché vanno tenute le chiavi esterne referenti all'id progressivo!! I dati all'interno della chiave creata possono essere nulli!

Reply:
? mhm no.

1 - la chiave esterna "id" non va bene
2 - la chiave esterna "unict_id" da sola non va bene perche' ci sono piu' valori uguali "unict_id". Ovvero, si possono verificare scenari in cui ci sono due `unict_id` uguali nello stesso `anno_accademico` che due unict_id uguali in diversi anni_accademici (ma questo e' logico).

mettendo queste altre chiavi la differenza e' assicurata, in particolare id_cds era fondamentale, perche' per esempio CRITTOGRAFIA La puoi incontrare sia in Informatica LM-18 (inf. magistrale) sia in Matematica LM-40 (mat. magistrale), perche' la materia e' IDENTICA, cambia solo l'id_cds, poi ci possono essere altri scenari in cui e' necessario mettere anche gli altri id.

Reply:
Semplicemente non puoi basare la chiave su valori che possono essere nulli